scriptworker/worker.py

```diff
@@ -90,8 +96,86 @@
+
+        Returns: status code of build
+
+        """
+        try:
+            # Note: claim_work(...) might not be safely interruptible! See
+            # https://bugzilla.mozilla.org/show_bug.cgi?id=1524069
+            tasks = await self._run_cancellable(claim_work(context))
+            if not tasks or not tasks.get('tasks', []):
+                await self._run_cancellable(asyncio.sleep(context.config['poll_interval']))
+                return None
+
+            # Assume only a single task, but should more than one fall through,
+            # run them sequentially.  A side effect is our return status will
+            # be the status of the final task run.
+            status = None
+            for task_defn in tasks.get('tasks', []):
+                prepare_to_run_task(context, task_defn)
+                reclaim_fut = context.event_loop.create_task(reclaim_task(context, context.task))
+                try:
+                    status = await do_run_task(context, self._run_cancellable, self._to_cancellable_process)
+                    artifacts_paths = filepaths_in_dir(context.config['artifact_dir'])
+                except WorkerShutdownDuringTask:
+                    shutdown_artifact_paths = [os.path.join('public', 'logs', log_file)
+                                               for log_file in ['chain_of_trust.log', 'live_backing.log']]
+                    artifacts_paths = [path for path in shutdown_artifact_paths
+                                       if os.path.isfile(os.path.join(context.config['artifact_dir'], path))]
+                    status = STATUSES['worker-shutdown']
+                status = worst_level(status, await do_upload(context, artifacts_paths))
+                await complete_task(context, status)
+                reclaim_fut.cancel()
+                cleanup(context)
+
+            return status
+
+        except asyncio.CancelledError:
+            return None
+
+    async def _run_cancellable(self, coroutine: types.coroutine):
+        self.future = asyncio.ensure_future(coroutine)
+        if self.is_cancelled:
+            self.future.cancel()
+        result = await self.future
+        self.future = None
+        return result
+
+    async def _to_cancellable_process(self, task_process: TaskProcess):
+        self.task_process = task_process
+
+        if self.is_cancelled:
+            await task_process.worker_shutdown_stop()
+
+        return task_process
+
+    async def cancel(self):
+        """Cancel current work."""
+        self.is_cancelled = True
+        if self.future is not None:
+            self.future.cancel()
+        if self.task_process is not None:
+            log.warning("Worker is shutting down, but a task is running. Terminating task")
+            await self.task_process.worker_shutdown_stop()
+
+
 # run_tasks {{{1
-async def run_tasks(context, creds_key="credentials"):
+async def run_tasks(context):
```

Comment:
I removed `creds_key` because it's [not being used in `run_tasks(...)`](https://github.com/mozilla-releng/scriptworker/blob/master/scriptworker/worker.py#L94-L132)

Reply:
https://github.com/mozilla-releng/scriptworker/pull/310#discussion_r256623365 . [`read_integration_creds`](https://github.com/mozilla-releng/scriptworker/blob/master/scriptworker/test/test_integration.py#L39) appears to do that for us. So... docstring update, and fix the tests, and then we're good :)